README.md
# uma.sh
このスクリプトはWindows10のwsl(ubuntu)で実行されることを想定して作られています。
それ以外の環境では動作を確認していません。

あるゲームのアセットのリネームや音声の抽出などができるスクリプトです。

## 使用する前に
/mnt/c/Users/ユーザ名/AppData/LocalLow/Cygames/umamusume
で実行してください(要はmetaがあるディレクトリ)。

これを実行する前にゲーム内でデータの一括ダウンロードは済ませておきましょう。
一度タイトル画面をタップして先に進まないとmetaファイルが更新されないのでそれもしておきましょう。

足りないコマンド等はスクリプト内でインストールできるようになっています。
ただしvgmstreamのインストール場所が気に入らないならそこは各々編集してください。

## 使用方法
```sh
./uma.sh
```
と入力し、エンターを押して実行すれば動きます。

### 使えるオプションについて
- -c : コピーするときにサイズを比較して等しい場合スキップします。
- -f : コピーするとき全てのファイルをコピーします。
- -j : コピーするときの並列処理数を30以下1以上の整数で設定できます。
あんまり数を増やすとパソコンが壊れないか心配なので上限を設けました。
20を超える数や数値以外が入力された場合1に設定し直されます。
- -p : 出力先のフォルダ名を変更できます(非推奨)。
ただしスペースを名前に含めることはできません。
- -s : 音声を変換するときの並行処理数を20以下1以上の整数で設定できます。
安全を見て上限を設けました。なにも設定していない場合音声の並行処理数はファイルコピーの並行処理数の1/2(切り捨て)になるように設定します。
20を超える数や数値以外が入力された場合もファイルコピーの並行処理数の1/2(切り捨て)になるように設定し直されます。
- -U : このスクリプト自体をアップデートします。
- -h : ヘルプを表示します。

例:
```sh
./uma.sh -c -j30 -s20
```

## 補足
環境構築(おまけ)の中身はZSHの環境を構築できるスクリプトです(bashのままだと少し使いにくいので)。
WSL入れたけど、どうすればいいかわからないって方はとりあえず実行してみるとある程度使える環境が出来上がると思います。

もしインストールや使用時にわからないことがあれば、解決に至る質問が先にでている可能性がありますので [issues](https://github.com/incompetence33/uma_sh/issues?q=is%3Aissue) を一度御覧ください。

それでも解決しない場合新しく質問を [new issues](https://github.com/incompetence33/uma_sh/issues/new/choose) からお書き下さい。


ChangeLog.txt
更新履歴

2021/07/29 19:30
大量にあったよくわからんsedをawkを使って実装することで高速化と可読性の向上を行ったつもり。
コメントも少しばかりつけておいた。
今までのやつから流用できるところは流用したけどなるべく０から作ったつもり。
動画の変換についてはFFmpegが必要で、人によっては最新ビルドを使いたい人や特定のコーデックを使いたい人もいそうなのでやめた。
while分のところは繰り返しが多いのでそのうち関数化して短縮するかもしれない。

2021/07/31  (ただの日記)
複数ストリームを出力しようと無理やりhexdump と xxd でawbのヘッダーと中身のhcaを全部結合させる方法で実装してたがvgmstreamの-sオプションで事足りることが分かったのでアップロードをやめた。

2021/08/01 2:50
複数ストリーム出力を実装できました。
私の環境では「音声のみ」で実行すると16分20秒かかりました。
awbファイルの大きさに対して出力が少ないなと思いつつ目をそらしてしまっていました。
本当に申し訳ございませんでした……

2021/08/02 5:56
175 ID:7Gw3/tWIさんより報告頂いたバグを修正しました。
具体的にはrmにfオプションをつけたのと(すべてではない)、metaからリストを作成する関数が呼ばれてなかったので呼ぶようにしました。

2021/08/02 21:29
sqlite3へのパイプでインデントしているとエラーが出てしまうのでechoの拡張で対応した。

2021/08/12 02:47
「ライブだけ」を選択したとき「sound/」にあるファイルをすべて処理してしまっていたので、修正しました。

2021/08/12 03:57
ライブのID表がうまく出力されていなかったので修正しました。

2021/08/19 07:03
ご指摘のあったようにvgmstreamのインストール部分の
git clone --depth 1
の--depth1を消しました対応が遅れて申し訳ない。

2021/08/20 15:11
音声を変換するときawbのトラック数とすでにあるwavファイルの数を比べてそれが等しいならそのawbは変換しないようにした。

2021/08/21 22:14
whileでdircp回すところを関数化しました。
その結果可読性が下がりましたすみません(「'"hoge"'」のようになっている文字列を$1に入れることができなかったため)。
コピーの挙動をオプションで変えることができるようにしました。「uma.sh -h」でご確認ください。
コピー時や音声の処理時にスキップした数も表示するようにしました。

2021/08/23 00:54
ライブID表が正しく出力されていない問題を修正しました。

2021/08/24 01:09 (ただの日記)
「vgmstream CLI decoder r1626」にて内包されているストリームをすべて書き出すオプションなどが追加されましたが、旧バージョンを使っている人もいると思うのでしばらくは従来のやり方で出力しようと思います。

2021/08/24 08:05 (ただの日記)
Ubuntuの環境をどう構築したらいいかわからない人向けのコーナーを作っておいた。
悪評だったら消す。
夏休みが終わったら更新頻度がさらに低下する可能性がある。

2021/08/25 23:21
vgmstreamのインストールにcmakeを採用することにした。
fzfのインストールはaptではなくgithubからダウンロードすることにしました。

2021/08/26 05:19
vgmstreamの表記揺れに対応した。

2021/08/26 11:40
インストール周りで相手の意思問わずインストールしてしまうようになっていたので修正した。
変数名の間違いを修正した。

2021/08/30 00:31
GitHubにリポジトリを作ったのでアップデート機能を付けてみた。
./uma.sh -U でアップデートできます。
その際、uma.shというシンボリックリンク作るので、カスタマイズしたuma.shを作って置いているなら先に移動しておいてください。

2021/09/03 11:01
オプションで並列コピーを可能にした。
詳しくは
./uma.sh -h
で確認してほしい。
ただ、並列コピーをするとバックグラウンドで実行するためかコピーされた数やスキップされた数がうまく表示されないのでそこは無視してください。

2021/10/01 17:03
vgmstreamのインストール方法を各々でビルドする方法から公式が配布しているバイナリをダウンロード形にした。
気軽にissueって話はどっか行ったんすかね？スレ無駄に消費しちゃってごめんね。

2021/10/20 23:00
metaファイルに変更があった場合ファイルをコピーするときのフラグを変更するようにしました。

2021/12/16 23:05
meta内の音声ファイルのナンバリングが0000からだったのでそれに揃えるようにした。
そのためすでに音声をawbからデコードしている場合同一のファイルが別名で出力されてしまうので一旦sound_wavフォルダを削除するほうがいいかもしれません。ご迷惑おかけします……

2022/02/05 03:49
環境構築用のスクリプトを改良しました。

2022/02/02 05:09
音声をデコードする処理を並列処理できるように改善しました(長らく待たせてすまんな)。
最大で20並行でできるようになっています。
詳細は
./uma.sh -h
を叩いて確認してください。
大体すべての音声を変換するのに今まで50分ほどかかっていましたが当方の環境では7分弱まで短縮されました。
並列コピーする際、カウントが正確にされるようになりました。
ただ、仕様上(特に音声デコードをするとき)並行処理のなかで処理が遅いやつがいるとカウントが急に巻き戻ることがありますが異常ではありません。
その他細かいブラッシュアップを行いました。
バグ等ありましたらissueしていただけるとありがたいです。

2022/03/26 18:33
ReadMeをわかりやすくした。
衣装を変えるスクリプトを一応置いておいた(実は前からあったけど更新したので)。使う人いるか知らんけど。

2022/03/28 04:05
ID表の出力方法を変えた。
学校始まっちゃうのでissueは来ててもすぐ返せないかもしれませんが気がついたらなるべく対応していきたいと思います。(バグ報告、機能改善案、ReadMe読んでも使い方がわからない箇所がある等)

2022/04/14 19:45
vgmstreamのダウンロード方法を追加した。
>>705 >>706 ID:SYSULcpz さんの言うようにファイルを分割しようかと考えましたが、Releasesからダウンロードしてそのまま使う人もいるようなのでそのたびに動かないと言われることを考えたら今のままでもいいのかと思いました。でもアドバイスは嬉しいです、ありがとうございます。
あとこのツールの不具合はissuesにお願いします……スレでツールの話し出すと怒られちゃうんで……

2022/04/15 22:29
ZshのSetupの簡略化を行った。
そのデバッグ中に自身の環境を上書きして死にました()。

2022/04/16 20:21
今までに何度か /mnt/c/Users/ユーザ名/AppData/LocalLow/Cygames/umamusume にcdしないで実行して動かないと言ってる方がいたのでuma.shの場所が正しければ動くように改良しました。

2022/04/16 20:42  (ただの日記)
>>706 ID:SYSULcpz さんの言うようにファイル分割するほうがいいとわかりつついい案が浮かばなかったのでやらなくていいかと思っていたが、やはりやったほうが良さそうなので手を付けることにした。

2022/07/13 07:48
衣装変更のスクリプトは問題があるようなので消した。
ファイルの分割は……しらないです…………

環境構築(等おまけ)/dress_change.sh
#!/bin/bash
LANG="ja_JP.UTF-8"
trap "echo;echo 'CTRL+C が入力されたので終了します';exit 1" SIGINT
if [[ ! $(basename "$(pwd -P)") == umamusume ]]; then echo "umamusumeフォルダーで実行してくれ";echo "/mnt/c/Users/ユーザ名/AppData/LocalLow/Cygames/umamusume";echo "が一般的だと思います。";exit 1;fi
if [[ -z meta ]]; then echo "mata ファイルがあるところで実行してくれ。";echo "/mnt/c/Users/ユーザ名/AppData/LocalLow/Cygames/umamusume";echo "が一般的だと思います。";echo "もしくは一度も起動していないためにmetaファイルがないという可能性もあります。";exit 1;fi

make_character_id_list(){
	echo -ne '.mode csv\nselect * from text_data;'| sqlite3 master/master.mdb | sed 's/\r//g' | grep -E '^6,' | awk -F, 'gsub("\"","",$4){printf "s/_%s_/%s/g\n" ,$3,$4}' > character_id_list.txt
}
make_dress_id_list(){
	make_character_id_list
	echo -ne '.mode csv\nselect * from text_data;'| sqlite3 master/master.mdb | sed 's/\r//g' | grep -E '^14,' | awk -F, 'BEGIN{printf "決定\n全部\n"}gsub("\"","",$4){if($3 ~ /^[0-9]{6}/){gsub("（.*）","",$4);if($3 ~ /^1/){i=0;j=4}if($3 ~ /^9/){i=3;j=6}if($3 !~ /(^900|^2)/){printf "%s__%s_@%s\n" ,$3,substr($3,i,j),$4}}else printf "%s_%s\n" ,$3,$4}' | sed -f character_id_list.txt > dress_id_list.txt
}
make_dress_list_after_change(){
	db_restore
	echo -ne '.mode csv\nselect * from dress_data;'| sqlite3 tmp_.db | sed 's/\r//g' > dress_list_after_change.txt
}
db_restore(){
	curl -s "$(echo 'aHR0cHM6Ly9wcmQtc3RvcmFnZS11bWFtdXN1bWUuYWthbWFpemVkLm5ldC9kbC9yZXNvdXJjZXMvR2VuZXJpYy8K' | base64 -d)$(echo -ne '.dump\n' | sqlite3 meta | sed 's/\r//g' | grep 'master.mdb.lz4' | awk -F",|'" '{printf "%s/%s",substr($12,0,2),$12}')" | lz4 -dc > tmp_.db
}
db_output(){
	rm -f output_db_.txt
	echo -ne ".output output_db_.txt \n.dump" | sqlite3 master/master.mdb
}
db_rewriting(){
	make_dress_id_list
	make_dress_list_after_change
	db_output
	i=0
	while true;do
		dress_before_id[${i}]="$(grep -E "^([0-9]{6}|決定|全部)" dress_id_list.txt | fzf --reverse --header="変換希望の衣装の選択" | awk -F'_' '{print $1}')"
		if [[ ${dress_before_id[${i}]} == "決定" ]]; then break;fi
	i=$((10#${i}+1))
	done
	dress_before_id=($(echo ${dress_before_id[@]} | tr ' ' '\n' | sort -n | uniq | sed '/^決定$/d'))
	if [[ ${#dress_before_id[@]} == 0 ]]; then echo "変化させる衣装を選ばなければいけません。";exit 1;fi
	if [[ "${dress_before_id[@]}" =~ 全部 ]]; then
		dress_after_id="$(grep -E "^[0-9]{1,5}_" dress_id_list.txt | fzf --reverse --header="変換後の衣装の選択" | awk -F'_' '{print $1}')"
		dress_before_id=($(awk -F'_' '{if($1 ~ /^[0-9]{6}/ && $1 !~ /^(900|2)/){print $1}}' dress_id_list.txt))
	elif [[ ${#dress_before_id[@]} -ge 2 ]]; then
		dress_after_id="$(grep -E "^[0-9]{1,5}_" dress_id_list.txt | fzf --reverse --header="変換後の衣装の選択" | awk -F'_' '{print $1}')"
	else
		if [[ ${dress_before_id} =~ ^90 ]]; then CHARA_ID="${dress_before_id:2}";else CHARA_ID="${dress_before_id:0:4}";fi
		dress_after_id="$(grep -E "^(90${CHARA_ID}|${CHARA_ID}|[0-9]{1,5}_)" dress_id_list.txt | fzf --reverse --header="変換後の衣装の選択" | awk -F'_' '{print $1}')"
	fi
	rm -f sed_tmp.txt
	for NUM in $(seq 0 $((10#${#dress_before_id[@]}-1)));do
		echo "$(grep "INSERT INTO dress_data" output_db_.txt | grep \(${dress_before_id[${NUM}]}),$(grep "^${dress_after_id}," dress_list_after_change.txt)" | awk -F, 'BEGIN{printf "s@^.*dress_data VALUES('${dress_before_id[${NUM}]}'.*$@"}{for(i=1;i<=25;i++){data=$(i);if(i ~ /(^2$|^6$|^7$|^8$|^9$|^10$|^15$|^16$|^18$|^23$|^24$)/){data=$(i+26)} printf "%s," ,data}}END{printf "%s@\n",$26}' >> sed_tmp.txt
	done
	sed -f sed_tmp.txt output_db_.txt | tee output_db_tmp.txt > /dev/null 2>&1
	rm -f master_.mdb;echo -ne ".read output_db_tmp.txt\n.exit"|sqlite3 master_.mdb && mv master_.mdb master/master.mdb
}


while getopts "rh" OPT;do
	case $OPT in
		"r" ) db_restore;mv tmp_.db master/master.mdb;exit 0;;
		* ) echo "衣装を変更するスクリプトです。"
		echo "勝負服と私服を変更することができます。"
		echo "options"
		echo "-r:"
		echo "	master.mdbを修復します。"
		echo "-h:"
		echo "	この文章を表示します。"
		exit 0;
	esac
done
db_rewriting
rm -f character_id_list.txt dress_id_list.txt tmp_.db output_db_.txt output_db_tmp.txt dress_list_after_change.txt sed_tmp.txt
